dhall/src/Dhall/TypeCheck.hs

```diff
@@ -937,86 +939,89 @@ infer typer = loop
+                    return _T₀'
 
-                                        Just (Just _) -> do
-                                            case Eval.toVHPi _T' of
-                                                Just (x, _A₀', _T₀') -> do
-                                                    return (_T₀' (fresh ctx x))
+                match y handler' (Just _A₁') =
+                    case Eval.toVHPi handler' of
+                        Just (x, _A₀', _T₀') -> do
+                            if Eval.conv values _A₀' _A₁'
+                                then do
+                                    let _T₁' = _T₀' (fresh ctx x)
 
-                                                Nothing -> do
-                                                    let _T'' = quote names _T'
+                                    let _T₁'' = quote names _T₁'
 
-                                                    die (HandlerNotAFunction y₀ _T'')
+                                    -- If x appears in _T₁'', it's index will be -1.
+                                    -- Any well-typed Var has a non-negative index,
+                                    -- so if we find a negative index, we know
+                                    -- something is wrong.
+                                    let badVar (Var (V _ n)) = n < 0
+                                        badVar _             = False
 
-                                return (Just y₀, _T₁')
+                                    if badVar _T₁'' || Lens.Family.anyOf Dhall.Core.subExpressions badVar _T₁''
+                                        then die (UnboundVariable x)
+                                        else return _T₁'
 
-            let _T₁'' = quote names _T₁'
+                                else do
+                                    let _A₀'' = quote names _A₀'
+                                    let _A₁'' = quote names _A₁'
 
-            _ <- loop ctx _T₁''
+                                    die (HandlerInputTypeMismatch y _A₁'' _A₀'')
 
-            let process y mU = do
-                    case Dhall.Map.lookup y yTs' of
                         Nothing -> do
-                            die (MissingHandler diffU)
+                            let handler'' = quote names handler'
 
-                        Just _T' -> do
-                            _T₃' <- do
-                                case mU of
-                                    Nothing -> do
-                                        return _T'
+                            die (HandlerNotAFunction y handler'')
 
-                                    Just _A₁' -> do
-                                        case Eval.toVHPi _T' of
-                                            Just (x, _A₀', _T₂') -> do
-                                                if Eval.conv values _A₀' _A₁'
-                                                    then do
-                                                        return ()
+            let matched =
```

Comment:
If you do:

```haskell
matched <- sequence (Data.Map.intersectionWithKey match (Dhall.Map.toMap yTs') (Dhall.Map.toMap yUs'))
```

... then you won't need to handle the `Either`s in `checkMatched`